### PR TITLE
API-key-rotation-update-api-key

### PR DESCRIPTION
**Pull request by GunaKKIBM**

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -3,8 +3,8 @@ module github.com/IBM/ibm-csi-common
 go 1.16
 
 require (
-	github.com/IBM/ibmcloud-volume-interface v1.0.0-beta8
-	github.com/IBM/ibmcloud-volume-vpc v1.0.0-beta12
+	github.com/IBM/ibmcloud-volume-interface v1.0.1-0.20211126135815-30d8a848c0e1
```
Comment (sameshai) [inline go.mod:6]: Update the version to release one once the interface changes are committed

**Review by sameshai — APPROVED** [review]
LGTM

**Review by sameshai — DISMISSED** [review]
Please change the release version in go.mod for interface library and fix the the travis. REST LGTM

**Review by ambiknai — APPROVED** [review]

**Review by skorati — APPROVED** [review]
LGTM

File: pkg/ibmcloudprovider/volume_provider.go [hunk 2]
```diff
@@ -138,21 +141,41 @@ func (icp *IBMCloudStorageProvider) GetProviderSession(ctx context.Context, logg
...
 	vpcBlockConfig := &vpcconfig.VPCBlockConfig{
 		VPCConfig:    icp.ProviderConfig.VPC,
 		IKSConfig:    icp.ProviderConfig.IKS,
 		APIConfig:    icp.ProviderConfig.API,
 		ServerConfig: icp.ProviderConfig.Server,
 	}
-	session, isFatal, err := provider_util.OpenProviderSessionWithContext(ctx, vpcBlockConfig, icp.Registry, icp.ProviderName, logger)
-	if err != nil || isFatal {
+
+	maxRetryAttempt := 3
```
Comment (GunaKKIBM) [inline pkg/ibmcloudprovider/volume_provider.go:158]: make this constant

**Comment by GunaKKIBM** [comment]
Test results:

Creating pvc by deleting api key, initially, the pvc is in pending state, kubernetes retries to create volume. Meanwhile, if the api key is reset, volume gets created. It won't be necessary to restart the pod. 
```
kubectl create -f pvc.yaml                                   
persistentvolumeclaim/csi-vpc-block-5iops-pvc-api-key-rotation40 created

gunakk@Gunas-MBP OPENSHIFT_VPC_CLUSTER_TESTING % kubectl describe pvc csi-vpc-block-5iops-pvc-api-key-rotation40 
Name:          csi-vpc-block-5iops-pvc-api-key-rotation40
Namespace:     default
StorageClass:  ibmc-vpc-block-5iops-tier
Status:        Bound
Volume:        pvc-d7da1591-4aec-474a-b5f2-a406fe4b815c
Labels:        <none>
Annotations:   pv.kubernetes.io/bind-completed: yes
               pv.kubernetes.io/bound-by-controller: yes
               volume.beta.kubernetes.io/storage-provisioner: vpc.block.csi.ibm.io
Finalizers:    [kubernetes.io/pvc-protection]
Capacity:      10Gi
Access Modes:  RWO
VolumeMode:    Filesystem
Used By:       <none>
Events:
  Type     Reason                 Age                   From                                                                                      Message
  ----     ------                 ----                  ----                                                                                      -------
  Warning  ProvisioningFailed     3m29s                 vpc.block.csi.ibm.io_ibm-vpc-block-csi-controller-0_1d3a9aa8-1bc4-4e29-90cc-1e07d9edfef1  failed to provision volume with StorageClass "ibmc-vpc-block-5iops-tier": rpc error: code = Internal desc = {RequestID: 9d2c7752-c1e4-491a-a61f-fbb45a30b7d7 , Code: InternalError, Description: Internal error occurred, BackendError: API key is not found, reset the api key. If reset, retry after few minutes, Action: Please check 'BackendError' tag for more details}
  Warning  ProvisioningFailed     2m29s                 vpc.block.csi.ibm.io_ibm-vpc-block-csi-controller-0_1d3a9aa8-1bc4-4e29-90cc-1e07d9edfef1  failed to provision volume with StorageClass "ibmc-vpc-block-5iops-tier": rpc error: code = Internal desc = {RequestID: 077e111a-e15c-46d4-a54e-0abc0cf919d7 , Code: InternalError, Description: Internal error occurred, BackendError: API key is not found, reset the api key. If reset, retry after few minutes, Action: Please check 'BackendError' tag for more details}
  Normal   Provisioning           88s (x4 over 4m29s)   vpc.block.csi.ibm.io_ibm-vpc-block-csi-controller-0_1d3a9aa8-1bc4-4e29-90cc-1e07d9edfef1  External provisioner is provisioning volume for claim "default/csi-vpc-block-5iops-pvc-api-key-rotation40"
  Warning  ProvisioningFailed     88s                   vpc.block.csi.ibm.io_ibm-vpc-block-csi-controller-0_1d3a9aa8-1bc4-4e29-90cc-1e07d9edfef1  failed to provision volume with StorageClass "ibmc-vpc-block-5iops-tier": rpc error: code = Internal desc = {RequestID: b1c7a84a-04ec-4f74-be16-47ef86202d18 , Code: InternalError, Description: Internal error occurred, BackendError: API key is not found, reset the api key. If reset, retry after few minutes, Action: Please check 'BackendError' tag for more details}
  Normal   ExternalProvisioning   20s (x18 over 4m29s)  persistentvolume-controller                                                               waiting for a volume to be created, either by external provisioner "vpc.block.csi.ibm.io" or manually created by system administrator
  Normal   ProvisioningSucceeded  11s                   vpc.block.csi.ibm.io_ibm-vpc-block-csi-controller-0_1d3a9aa8-1bc4-4e29-90cc-1e07d9edfef1  Successfully provisioned volume pvc-d7da1591-4aec-474a-b5f2-a406fe4b815c

 % kubectl create -f validPOD2.yaml
deployment.apps/api-key-rotation-pod40 created

 % kubectl get pods
NAME                                      READY   STATUS             RESTARTS   AGE
api-key-rotation-pod40-6fff7bb57-vmnxn    1/1     Running            0          3m26s
```

File: go.sum [hunk 3]
```diff
@@ -4,10 +4,10 @@ github.com/BurntSushi/toml v0.3.1 h1:WXkYYl6Yr3qBf1K79EBnL4mak0OimBfB0XUf9Vl28OQ
 github.com/BurntSushi/toml v0.3.1/go.mod h1:xHWCNGjB5oqiDr8zfno3MHue2Ht5sIBksp03qcyfWMU=
 github.com/IBM-Cloud/ibm-cloud-cli-sdk v0.6.7 h1:eHgfQl6IeSmzWUyiSi13CvoFYsovoyqWlpHX0pa9J54=
 github.com/IBM-Cloud/ibm-cloud-cli-sdk v0.6.7/go.mod h1:RiUvKuHKTBmBApDMUQzBL14pQUGKcx/IioKQPIcRQjs=
-github.com/IBM/ibmcloud-volume-interface v1.0.0-beta8 h1:0oCBQidejAlLI5f5PlBHIytArhJI1aEOBCCOMoxNOPs=
-github.com/IBM/ibmcloud-volume-interface v1.0.0-beta8/go.mod h1:2LxqkERml3eBxfRkQniuiRmKkAnl5tfC0ttlP2gupJM=
-github.com/IBM/ibmcloud-volume-vpc v1.0.0-beta12 h1:Hfw7XvHpexffDfvD50wFJCwUmhKeVUlKEFi7JMcSSZI=
-github.com/IBM/ibmcloud-volume-vpc v1.0.0-beta12/go.mod h1:MEyNF/zxHl2Zw/kBy7qykGPt8X+U+kDEJV1kimrvCbs=
+github.com/IBM/ibmcloud-volume-interface v1.0.1-0.20211203045519-74fe99e68e4f h1:XcW0pted1N9FWNhtspx38USjCmWduICvngyeU/OZ34g=
```
Comment (arahamad) [inline go.sum:7]: need to be updated for v1.0.0-beta9 release

File: pkg/ibmcloudprovider/volume_provider.go [hunk 4]
```diff
@@ -138,21 +143,45 @@ func (icp *IBMCloudStorageProvider) GetProviderSession(ctx context.Context, logg
...
+		session, _, err := provider_util.OpenProviderSessionWithContext(ctx, prov, vpcBlockConfig, icp.ProviderName, logger)
+		if err == nil {
+			logger.Info("Successfully got the provider session", zap.Reflect("ProviderName", session.ProviderName()))
+			return session, nil
+		}
 		logger.Error("Failed to get provider session", zap.Reflect("Error", err))
+		if retryCount == 1 {
+			return nil, err
+		}
+		if providerError, ok := err.(provider.Error); ok && providerError.Code() == reasoncode.ErrorFailedTokenExchange && (strings.Contains(strings.ToLower(providerError.Error()), "api key could not be found") || strings.Contains(strings.ToLower(providerError.Error()), "user not found or active")) {
```
Comment (arahamad) [inline pkg/ibmcloudprovider/volume_provider.go:170]: can we keep these hardcoded strings in some const file or variable

File: pkg/ibmcloudprovider/volume_provider.go [hunk 5]
```diff
@@ -164,3 +193,67 @@ func (icp *IBMCloudStorageProvider) GetConfig() *config.Config {
 func (icp *IBMCloudStorageProvider) GetClusterInfo() *utils.ClusterInfo {
 	return icp.ClusterInfo
 }
+
+// UpdateAPIKey ...
+func (icp *IBMCloudStorageProvider) UpdateAPIKey(logger *zap.Logger) error {
+	logger.Info("Updating API key in cloud storage provider")
+	vpcBlockConfig := &vpcconfig.VPCBlockConfig{
```
Comment (arahamad) [inline pkg/ibmcloudprovider/volume_provider.go:200]: add the comments which will explain the logic

File: pkg/ibmcloudprovider/volume_provider.go [hunk 6]
```diff
@@ -164,3 +193,67 @@ func (icp *IBMCloudStorageProvider) GetConfig() *config.Config {
...
+		VPCConfig:    icp.ProviderConfig.VPC,
+		IKSConfig:    icp.ProviderConfig.IKS,
+		APIConfig:    icp.ProviderConfig.API,
+		ServerConfig: icp.ProviderConfig.Server,
+	}
+	vpcAPIKey := vpcBlockConfig.VPCConfig.G2APIKey
```
Comment (arahamad) [inline pkg/ibmcloudprovider/volume_provider.go:206]: why we need this,pl add comments

File: pkg/ibmcloudprovider/volume_provider.go [hunk 7]
```diff
@@ -164,3 +193,67 @@ func (icp *IBMCloudStorageProvider) GetConfig() *config.Config {
...
+		logger.Info("Created NewAPIKeyImpl...")
+		err = apiKeyImp.UpdateIAMKeys(icp.ProviderConfig)
+		if err != nil {
+			logger.Error("Unable to get API key", local.ZapError(err))
+			return err
+		}
+		// If the retrieved API key is the same as previous one, returning error
+		if vpcAPIKey == icp.ProviderConfig.VPC.G2APIKey {
+			logger.Error("API key is not reset")
+			return errors.New("API key is not found, reset the api key. If reset, retry after few minutes")
```
Comment (arahamad) [inline pkg/ibmcloudprovider/volume_provider.go:223]: keep this error msg https://github.com/IBM/ibm-csi-common/blob/master/pkg/messages/messages_en.go and use it

File: pkg/ibmcloudprovider/volume_provider.go [hunk 8]
```diff
@@ -164,3 +193,67 @@ func (icp *IBMCloudStorageProvider) GetConfig() *config.Config {
...
+		icp.ProviderConfig.VPC.APIKey = conf.VPC.APIKey
+		icp.ProviderConfig.VPC.G2APIKey = conf.VPC.G2APIKey
+		vpcBlockConfig.VPCConfig.APIKey = conf.VPC.G2APIKey
+		vpcBlockConfig.VPCConfig.G2APIKey = conf.VPC.G2APIKey
+	}
+
+	prov, err := icp.Registry.Get(icp.ProviderName)
+	if err != nil {
+		logger.Error("Not able to get the said provider, it might not registered", local.ZapError(err))
+		return errors.New("unable to fetch the provider, hence unable to update the new API key")
```
Comment (arahamad) [inline pkg/ibmcloudprovider/volume_provider.go:249]: pl keep this in message_en.go file

File: pkg/ibmcloudprovider/volume_provider.go [hunk 9]
```diff
@@ -164,3 +193,67 @@ func (icp *IBMCloudStorageProvider) GetConfig() *config.Config {
...
+	}
+
+	err = prov.UpdateAPIKey(vpcBlockConfig, logger)
+	if err != nil {
+		logger.Error("Failed to update API key in the provider", local.ZapError(err))
+		return errors.New("unable to update the new API key")
```
Comment (arahamad) [inline pkg/ibmcloudprovider/volume_provider.go:255]: if it will be seen by user pl add in message_en.go file

**Review by arahamad — CHANGES_REQUESTED** [review]
pl check comments

**Review by arahamad — APPROVED** [review]
lgtm

File: go.mod (hunk 1 above)
Comment (arahamad) [reply go.mod]: done

File: pkg/ibmcloudprovider/volume_provider.go (hunk 6 above)
Comment (GunaKKIBM) [reply pkg/ibmcloudprovider/volume_provider.go:216]: Added

**Review by prankulmahajan — APPROVED** [review]
lgtm

File: pkg/ibmcloudprovider/volume_provider.go (hunk 5 above)
Comment (GunaKKIBM) [reply pkg/ibmcloudprovider/volume_provider.go:208]: Done

File: pkg/ibmcloudprovider/volume_provider.go (hunk 4 above)
Comment (GunaKKIBM) [reply pkg/ibmcloudprovider/volume_provider.go]: Done

File: go.sum (hunk 3 above)
Comment (GunaKKIBM) [reply go.sum]: Done

File: pkg/ibmcloudprovider/volume_provider.go (hunk 2 above)
Comment (GunaKKIBM) [reply pkg/ibmcloudprovider/volume_provider.go]: Done

**Pull request merged**